zkfacade/src/main/java/com/yahoo/vespa/curator/Curator.java

```diff
@@ -47,96 +47,105 @@
     private static final int ZK_SESSION_TIMEOUT = 30000;
     private static final int ZK_CONNECTION_TIMEOUT = 30000;
 
-    private static final int baseSleepTime = 1000; //ms
-    private static final int maxRetries = 10;
+    private static final int BASE_SLEEP_TIME = 1000; //ms
+    private static final int MAX_RETRIES = 10;
 
-    private final CuratorFramework curatorFramework;
     protected final RetryPolicy retryPolicy;
 
-    private final String connectionSpec;
-    private final int serverCount;
+    private final CuratorFramework curatorFramework;
+    private final String connectionSpec; // May be a subset of the servers in the ensemble
+
+    private final String zooKeeperEnsembleConnectionSpec;
+    private final int zooKeeperEnsembleCount;
 
     /** Creates a curator instance from a comma-separated string of ZooKeeper host:port strings */
     public static Curator create(String connectionSpec) {
-        return new Curator(connectionSpec);
+        return new Curator(connectionSpec, connectionSpec);
     }
 
     // Depend on ZooKeeperServer to make sure it is started first
     // TODO: Move zookeeperserver config out of configserverconfig (requires update of controller services.xml as well)
     @Inject
     public Curator(ConfigserverConfig configserverConfig, ZooKeeperServer server) {
-        this(createConnectionSpec(configserverConfig));
+        this(configserverConfig, createConnectionSpec(configserverConfig));
     }
-    
-    static String createConnectionSpec(ConfigserverConfig config) {
-        String thisServer = HostName.getLocalhost();
-
-        StringBuilder sb = new StringBuilder();
-        for (int i = 0; i < config.zookeeperserver().size(); i++) {
-            ConfigserverConfig.Zookeeperserver server = config.zookeeperserver(i);
-
-            String spec = String.format("%s:%d", server.hostname(), server.port());
 
-            if (config.zookeeperLocalhostAffinity() && server.hostname().equals(thisServer)) {
-                // Only connect to localhost server if possible, to save network traffic
-                // and balance load.
-                return spec;
-            }
+    private Curator(ConfigserverConfig configserverConfig, String zooKeeperEnsembleConnectionSpec) {
+        this(configserverConfig.zookeeperLocalhostAffinity() ?
+                HostName.getLocalhost() : zooKeeperEnsembleConnectionSpec,
```

Comment:
Dont you need ":port" as well when configserverConfig.zookeeperLocalhostAffinity()  is true?

Reply:
Yes! Good catch, fix coming up.